packages/@aws-cdk/aws-stepfunctions/lib/fields.ts

```diff
@@ -0,0 +1,136 @@
+   */
+  public static fromListAt(path: string): string[] {
+    validateDataPath(path);
+    return new JsonPathToken(path).toList();
+  }
+
+  /**
+   * Instead of using a literal number, get the value from a JSON path
+   */
+  public static fromNumberAt(path: string): number {
+    validateDataPath(path);
+    return new JsonPathToken(path).toNumber();
+  }
+
+  /**
+   * Use the entire data structure
+   *
+   * Will be an object at invocation time, but is represented in the CDK
+   * application as a string.
+   */
+  public static get entirePayload(): string {
+    return new JsonPathToken('$').toString();
+  }
+
+  private constructor() {
+  }
+}
+
+/**
+ * Extract a field from the State Machine Context data
+ *
+ * @see https://docs.aws.amazon.com/step-functions/latest/dg/connect-to-resource.html#wait-token-contextobject
+ */
+export class ContextField {
+  /**
+   * Instead of using a literal string, get the value from a JSON path
+   */
+  public static fromStringAt(path: string): string {
+    validateContextPath(path);
+    return new JsonPathToken(path).toString();
+  }
+
+  /**
+   * Instead of using a literal number, get the value from a JSON path
+   */
+  public static fromNumberAt(path: string): number {
+    validateContextPath(path);
+    return new JsonPathToken(path).toNumber();
+  }
+
+  /**
+   * Return the Task Token field
+   *
+   * External actions will need this token to report step completion
+   * back to StepFunctions using the `SendTaskSuccess` or `SendTaskFailure`
+   * calls.
+   */
+  public static get taskToken(): string {
+    return new JsonPathToken('$$.Task.Token').toString();
+  }
+
+  /**
+   * Use the entire context data structure
+   *
+   * Will be an object at invocation time, but is represented in the CDK
+   * application as a string.
+   */
+  public static get entireContext(): string {
+    return new JsonPathToken('$$').toString();
+  }
+
+  private constructor() {
+  }
+}
+
+/**gg
+ * Helper functions to work with structures containing fields
+ */
+export class FieldUtils {
+
+  /**
+   * Render a JSON structure containing fields to the right StepFunctions structure
+   */
+  public static renderObject(obj?: {[key: string]: any}): {[key: string]: any}  | undefined {
+    return renderObject(obj);
+  }
+
+  /**
+   * Return all JSON paths used in the given structure
+   */
+  public static findReferencedPaths(obj?: {[key: string]: any}): string[] {
+    return Array.from(findReferencedPaths(obj)).sort();
+  }
+
+  /**
+   * Returns whether the given task structure contains the TaskToken field anywhere
+   *
+   * The field is considered included if the field itself or one of its containing
+   * fields occurs anywhere in the payload.
+   */
+  public static containsTaskToken(obj?: {[key: string]: any}): boolean {
+    const paths = findReferencedPaths(obj);
+    return paths.has('$$.Task.Token') || paths.has('$$.Task') || paths.has('$$');
```

Comment:
I don't understand the idea here. As far as I know, it is mandatory for users to specify "$$.Task.Token" in the top level of "Payload".

Reply:
Mandated by whom? Will SFN reject the state machine definition if the token is not contained?

I'm asking because given `$$.Task` or `$$`, a user could still get to the value and send it in `SendTaskSuccess`. But if SFN's API will actively reject a a StateMachine definition that doesn't contain that field, then we have to cater to the strictest validation.

Reply:
Also, re: top-level, will SFN reject the following state machine definition:

```
"Parameters": {
  "MessageBody": {
    "Input": "Some String",
    "MoreData": {
      "Something": "$.Field",
      "Token": "$$.Task.Token"
    }
  }
}
```

Reply:
Thank you for your explanation. I understand better now.

To correct my mistake: 
Users can include "$$.Task.Token" anywhere under "Parameters". To pass this variable to the target service, the token should be included in the subfield "Payload" for Lambda, "MessageBody" for SQS, "Message" for SNS. So your example works:
```
"Parameters": {
  "MessageBody": {
    "Input": "Some String",
    "MoreData": {
      "Something.$": "$.Field",
      "Token.$": "$$.Task.Token"
    }
  }
}
```
In addition, I checked in the code as well as tested myself, Step Functions does not reject if the token field is missing for callback(.waitForTaskToken) tasks.

Reply:
Cool. I will encourage the validation in the CDK, because we're all about early validation and early feedback here. But at least we can set up our own best practices then.